documentation/specs/proposed/WASI-tasks.md

```diff
@@ -0,0 +1,182 @@
+An important criterion is popularity/activity in development as the WASM standard is evolving and needs a lot of developers to implement it.
+This leads to considering [Wasmtime](https://wasmtime.dev/) or [Wasmer](https://wasmer.io/).
+Interaction with C# is especially important for us so we will use **Wasmtime** because the integration via a NuGet package is more up to date and there is more active development in tooling and other dotnet projects use it. 
+
+### Testing
+- **TBD**
+- E2E tests - building projects in different languages most important
+
+<!-- Integration tests for logging -->
+<!-- mirror MSBuild\src\Build.UnitTests\BackEnd\TaskHost_Tests.cs --> 
+
+## User Experience
+API should be clear and the Rust task provide an example of how to implement it.
+Then the user adds the task to their .proj file and it runs and logs as if it were a C# task.
+
+## Development remarks (in-progress)
+
+### TODO for this doc
+- create in depth explanations for Wasm/WASI and how its concepts map to MSBuild concepts
+- discuss with people who understand MSBuild internals, WASI and dotnet interaction, users of MSBuild
+
+### Open questions and considerations 
+- **implementing WASI api on our own like [wasm in vscode](https://github.com/microsoft/vscode-wasm)?**
+    - security👍 we don't have control over external runtime
+    - customizable👍
+    - hard to maintain👎, wasi is constantly changing
+    - lot of work 👎
+
+
+- **Interacting with the tooling for creating .wasi files**
+    - hard, unstable
+    - separate project?
+    - out of scope?
+    - if yes - inclusion in msbuild? how to handle missing?
+
+- **bundling wasm runtime with MSBuild?**
+    - compatibility👍
+    - ease of use 👍
+    - size👎
+    - maintenance👎
+    - if no how to handle missing?
+
+#### Related issues making other environments for running tasks:
+https://github.com/dotnet/msbuild/issues/711
+https://github.com/dotnet/msbuild/issues/4834
+https://github.com/dotnet/msbuild/issues/7257
+
+### Requirements gathering
+What are the needs that WASI tasks would address?
+WASI task host? security?
+
+
+### Related projects
+
+[wasmtime](https://wasmtime.dev/) - Wasm runtime
+
+[wasm in vscode](https://github.com/microsoft/vscode-wasm)
+
+[wasmtime-dotnet](https://github.com/bytecodealliance/wasmtime-dotnet)
+
+[dotnet-wasi-sdk](https://github.com/dotnet/dotnet-wasi-sdk) - compile dotnet to Wasm - likely stale
+- copy their properties as those would be similar
+
+[wasm in dotnet runtime](https://github.com/dotnet/runtime/tree/main/src/mono/wasm)
+https://github.com/dotnet/runtime/discussions/98538#discussioncomment-8499105
+https://github.com/dotnet/runtime/issues/65895#issuecomment-1511265657
+
+[componentize-dotnet](https://github.com/bytecodealliance/componentize-dotnet) nuget package to make Wasi bundle from c#, released a week ago
+
+### Random
+System.Runtime.InteropServices
+
+windows/linux/mac all seem like they will have some specifics
+
+threadsafety?
```

Comment:
Ah yeah good point, there are two sides of thread safety:

1. _Inside_ the task
2. In MSBuild _outside_ the task.

Probably fine to deprioritize both, inside for @baronfel's reason and outside because task execution should appear single-threaded (due to MSBuild's multi-process instead of multi-thread model).